File: Photino/EUMCommunication/FileSystem/FileSystemCommunication.cs
using EUM.Core.Serialization;
using HelloPhotino.Vue.Communication;
using HelloPhotino.Vue.FileSystem;
using Photino.NET;

namespace EUM.Photino.EUMCommunication.FileSystem
{
    internal static class FileSystemCommunication
    {
        internal static ICommunicationResponse ReadDir(ICommunicationRequest request, PhotinoWindow _)
        {
            var response = new BaseCommunicationResponse(LayerCommunicationResult.NOK, request.CallerName, request.RequestId);
            var payload = JsonSerializerExtender.Deserialize<ReadDirPayload>(request.Message);
            if (payload == null)
            {
                return response;
            }


            var entries = ProcessDir(payload.Path);

            response.Message = JsonSerializerExtender.Serialize(entries);
            response.Result = LayerCommunicationResult.OK;
            return response;
        }

        internal static List<FileSystemEntry> ProcessDir(string path)
        {
            var entries = new List<FileSystemEntry>();
            var dirInfo = new DirectoryInfo(path);
            if (!dirInfo.Exists)
            {
                return entries;
            }

            foreach (var entry in dirInfo.EnumerateFileSystemInfos())
            {
                if (entry is null)
                {
                    continue;
                }

                var isDir = (entry.Attributes & FileAttributes.Directory) == FileAttributes.Directory;

                var fileSystemEntry = new FileSystemEntry
                {
                    Name = entry.Name,
                    FullPath = entry.FullName,
                    Extension = isDir ? string.Empty : entry.Extension,
                    PathWithoutFileName = entry.FullName ?? "",
                    Size = isDir ? 0 : ((FileInfo)entry).Length,
                    LastModified = new DateTimeOffset(entry.LastWriteTimeUtc).ToUnixTimeSeconds(),
                    IsDirectory = isDir
                };

                if (isDir)
                {
                    fileSystemEntry.Children = ProcessDir(entry.FullName);
                }

                entries.Add(fileSystemEntry);
            }

            return entries;
        }

        internal static ICommunicationResponse DeleteFile(ICommunicationRequest request, PhotinoWindow window)
        {
            var response = new BaseCommunicationResponse(LayerCommunicationResult.NOK, request.CallerName, request.RequestId);
            if (File.Exists(request.Message))
            {
                File.Delete(request.Message);
                response.Result = LayerCommunicationResult.OK;
            }

            return response;
        }

        internal static async Task<ICommunicationResponse> OpenFileDialogAsync(ICommunicationRequest request, PhotinoWindow window)
        {
            var response = new BaseCommunicationResponse(LayerCommunicationResult.NOK, request.CallerName, request.RequestId);
            var payload = JsonSerializerExtender.Deserialize<OpenFileDialogPayload>(request.Message);
            if (payload == null)
            {
                return response;
            }

            if (payload.SelectFolder)
            {
                var selectedFolders = await window.ShowOpenFolderAsync(defaultPath: payload.DefaultPath, multiSelect: payload.MultiSelect);
                var folderResponse = new OpenFileDialogResponse
                {
                    SelectedPaths = selectedFolders
                };

                response.Result = LayerCommunicationResult.OK;
                response.Message = JsonSerializerExtender.Serialize(folderResponse);
                return response;
            }

            (string Name, string[] Extensions)[]? filters = null;
            if (payload.AllowedExtension?.Length > 0)
            {
                filters =
                [
                    (payload.FilterName, payload.AllowedExtension)
                ];
            }

            var defaultPath = string.Empty;
            if (!string.IsNullOrEmpty(payload.DefaultPath))
            {
                defaultPath = payload.DefaultPath.Replace('/', Path.DirectorySeparatorChar);
            }

            var selectedFiles = await window.ShowOpenFileAsync(defaultPath: defaultPath, multiSelect: payload.MultiSelect, filters: filters);
            var fileResponse = new OpenFileDialogResponse
            {
                SelectedPaths = selectedFiles
            };

            response.Result = LayerCommunicationResult.OK;
            response.Message = JsonSerializerExtender.Serialize(fileResponse);
            return response;
        }

        internal static ICommunicationResponse GetUserDocumentFolder(ICommunicationRequest request, PhotinoWindow _1)
        {
            var appDocumentPath = BaseFileSystemProvider.GetAppDocumentPath();
            return new BaseCommunicationResponse(LayerCommunicationResult.OK, request.CallerName, request.RequestId)
            {
                Message = appDocumentPath
            };
        }

        internal static ICommunicationResponse GetFileMetadata(ICommunicationRequest request, PhotinoWindow _1)
        {
            var response = new BaseCommunicationResponse(LayerCommunicationResult.OK, request.CallerName, request.RequestId);
            var filePath = request.Message;

            var getFileMetadaResponse = new GetFileMetadataResponse
            {
                LastModifiedTimestamp = 0,
            };
            if (!File.Exists(filePath))
            {
                response.Message = JsonSerializerExtender.Serialize(getFileMetadaResponse);
                return response;
            }

            var fileInfo = new FileInfo(filePath);
            if (fileInfo is null)
            {
                response.Message = JsonSerializerExtender.Serialize(getFileMetadaResponse);
                return response;
            }

            getFileMetadaResponse.LastModifiedTimestamp = ((DateTimeOffset)fileInfo.LastWriteTimeUtc).ToUnixTimeSeconds();

            response.Message = JsonSerializerExtender.Serialize(getFileMetadaResponse);
            response.Result = LayerCommunicationResult.OK;

            return response;
        }
    }
}

File: Photino/UserInterface/src/scripts/layerCommunication/fileSystemCommunication.ts
import { dataDir, saveDataDir } from "../appContext";
import { replaceAllCharaters } from "../utils";
import { LayerCommunicationResult, sendRequest, type BaseCommunicationRequest } from "./baseLayerCommunication";

export async function readDir(path: string) : Promise<FileSystemEntry[]> {
    var requestMessage = JSON.stringify({ path });
    var request = <BaseCommunicationRequest>({ callerName: 'ReadDir', message: requestMessage });
    var response = await sendRequest<FileSystemEntry[]>(request);

    if(response.hasPayload) {
        return response.payload;
    }

    return new Array();
}

export async function openFileDialog(parameters: openFileDialogRequest) : Promise<string[] | undefined> {
    var requestMessage = JSON.stringify(parameters);
    var request = <BaseCommunicationRequest>({ callerName: 'OpenFileDialogAsync', message: requestMessage });
    var response = await sendRequest<openFileDialogResponse>(request);

    return response.payload?.selectedPaths;
}

export async function deleteFile(filePath: string) {
    var request = <BaseCommunicationRequest>({ callerName: 'DeleteFile', message: filePath });
    await sendRequest(request);
}

export async function getUserDocumentFolder() : Promise<string | undefined> {
    if(!dataDir || dataDir === '' || dataDir === '/') {
        var request = <BaseCommunicationRequest>({ callerName: 'GetUserDocumentFolder' });
        var response = await sendRequest<openFileDialogResponse>(request);
        
        if(response.result == LayerCommunicationResult.NOK) {
            return;
        }

        saveDataDir(replaceAllCharaters(response.message, '\\\\', '/').trimEnd());
    }

    return dataDir;
}

export async function getFileMetadata(filePath: string) : Promise<GetFileMetadataResponse | undefined> {
    
    var request = <BaseCommunicationRequest>({ callerName: 'GetFileMetadata', message: filePath });
    var response = await sendRequest<GetFileMetadataResponse>(request);
        
    if(!response || response.result == LayerCommunicationResult.NOK) {
        return;
    }
    
    return response.payload;
}

export type openFileDialogRequest = {
    defaultPath: string | undefined,
    multiSelect: boolean,
    selectFolder: boolean,
    filterName: string | undefined,
    allowedExtension: string[] | undefined
}

export type openFileDialogResponse = {
    selectedPaths: string[] | undefined
}

export type GetFileMetadataResponse = {
    lastModifiedTimestamp: number
}

export type FileSystemEntry = {
    name: string,
    fullPath: string,
    extension: string,
    pathWithoutFileName: string,
    size: number,
    lastModifed: number,
    isDirectory: boolean,
    children: FileSystemEntry[]
}